### ISSUE TITLE
[strict-type-predicates] false positive warning

### ISSUE THREAD

**Issue by zheeeng** [palantir/tslint#4238]
### Bug Report

- __TSLint version__: 5.11.0
- __TypeScript version__: .1.3
- __Running TSLint via__: create-react-app-typescript

#### TypeScript code being linted

TS report: `Expression is always true.`

```ts
export function test <O extends object> (o: O) {
  for (const key in o) {
    if (o.hasOwnProperty(key)) {
      const condition = typeof merged[key] !== 'object'  // Expression is always true.

      console.log(condition)
    }
  }
}
```

TS report: `Expression is always false.`

```ts
export function test <O extends object> (o: O) {
  for (const key in o) {
    if (o.hasOwnProperty(key)) {
      const condition = typeof merged[key] !== 'number'  // Expression is always false.

      console.log(condition)
    }
  }
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "strict-type-predicates": true
  }
}
```

#### Actual behavior

console warns.

#### Expected behavior

Two snippets above all pass.

**Comment by JoshuaKGoldberg** [comment]
@zheeeng what type is `merged`? The `strict-type-predicates` rule needs that information, so we can't really help you solve this issue without it.

Is this a duplicate of #2944?

**Comment by zheeeng** [comment]
My bad, my original codes are:

```
export default function mergeOptions<O> (...options: O[]) {

  return options.reduce(
    (merged, option) => {
      for (const key in option) {
        if (!option.hasOwnProperty(key)) continue

        merged[key] = (
          // tslint:disable-next-line:strict-type-predicates
          !merged[key] || typeof merged[key] !== 'object' || [merged[key], option[key]].some(Array.isArray)
        )
          ? option[key]
          : mergeOptions(merged[key] as any, option[key] as any)
      }

      return merged
    },
    // tslint:disable-next-line:no-object-literal-type-assertion
    {} as O,
  )
}
```
But after removing the commnet, these codes run fine.
